Drop CString::mutableData(), in favor of mutableSpan()

cdumez opened the pull request:
#### 837e13b4365be7092a7f31e63e271ebf4beeb608
<pre>
Drop CString::mutableData(), in favor of mutableSpan()
<a href="https://bugs.webkit.org/show_bug.cgi?id=284580">https://bugs.webkit.org/show_bug.cgi?id=284580</a>

Reviewed by Darin Adler.

* Source/JavaScriptCore/runtime/IntlDisplayNames.cpp:
(JSC::IntlDisplayNames::of const):
* Source/WTF/wtf/cocoa/FileSystemCocoa.mm:
(WTF::FileSystemImpl::createTemporaryZipArchive):
* Source/WTF/wtf/darwin/OSLogPrintStream.mm:
(WTF::OSLogPrintStream::vprintf):
* Source/WTF/wtf/posix/FileSystemPOSIX.cpp:
(WTF::FileSystemImpl::makeAllDirectories):
* Source/WTF/wtf/text/CString.cpp:
(WTF::CStringBuffer::createUninitialized):
(WTF::CString::copyBufferIfNeeded):
(WTF::CString::grow):
(WTF::CString::mutableData): Deleted.
* Source/WTF/wtf/text/CString.h:
(WTF::CString::data const):
(WTF::CString::spanIncludingNullTerminator const):
* Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
(WebKit::temporaryPDFDirectoryPath):
(WebKit::pathToPDFOnDisk):
* Tools/TestWebKitAPI/Tests/WTF/CString.cpp:
(TEST(WTF, CStringCopyOnWrite)):
* Tools/TestWebKitAPI/Tests/WTF/SynchronizedFixedQueue.cpp:
(TestWebKitAPI::toUpper):

Canonical link: <a href="https://commits.webkit.org/287795@main">https://commits.webkit.org/287795@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3ed6ac510053b63a1416fb05d53f4a15e1556c50

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/80823 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/345 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/34759 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/85350 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/31807 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/82934 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/362 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/8144 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/63112 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20893 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/83892 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/173 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/73564 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43416 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/95 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/27724 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/30264 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/73803 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/71656 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/28258 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/86782 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/79882 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/8050 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/5679 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/71419 "Found 15 new test failures: imported/w3c/web-platform-tests/css/css-transforms/animation/transform-interpolation-scale.html imported/w3c/web-platform-tests/css/css-transforms/animation/transform-interpolation-translate-em.html imported/w3c/web-platform-tests/css/css-transforms/animation/translate-percent-with-width-and-height-separate.html imported/w3c/web-platform-tests/css/css-transitions/pseudo-element-transform.html imported/w3c/web-platform-tests/css/css-values/vh-interpolate-pct.html imported/w3c/web-platform-tests/css/css-view-transitions/active-view-transition-type-on-non-root.html imported/w3c/web-platform-tests/css/css-view-transitions/animating-new-content.html imported/w3c/web-platform-tests/css/css-view-transitions/auto-name.html imported/w3c/web-platform-tests/css/css-view-transitions/old-content-with-overflow.html imported/w3c/web-platform-tests/web-animations/timing-model/animations/infinite-duration-animation.html ... (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/8227 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/69400 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70660 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14677 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/13616 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/102289 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12539 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/8012 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/13533 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/24872 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/7851 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/11370 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/9657 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/JavaScriptCore/runtime/IntlDisplayNames.cpp:
```diff
@@ -156,19 +156,20 @@ JSValue IntlDisplayNames::of(JSGlobalObject* globalObject, JSValue codeValue) co
-            char* mutableData = result.mutableData();
-            for (unsigned index = 0; index < result.length(); ++index)
-                mutableData[index] = toASCIIUpper(mutableData[index]);
+            auto mutableData = result.mutableSpan();
```
We don't need this local variable.

darinadler commented on Source/WTF/wtf/darwin/OSLogPrintStream.mm:
```diff
@@ -58,12 +58,12 @@
     va_copy(backup, argList);
 ALLOW_NONLITERAL_FORMAT_BEGIN
-    size_t bytesWritten = vsnprintf(m_string.mutableData() + offset, freeBytes, format, argList);
+    size_t bytesWritten = vsnprintf(m_string.mutableSpanIncludingNullTerminator().subspan(offset).data(), freeBytes, format, argList);
```
Too bad this will not detect if `offset` is past the null character.

kmiller68 replied:
Does `subspan` (with release asserts) not check the bounds? It's UB to create an out of bounds `subspan` so crashing would be valid. Maybe we could request this?

cdumez replied:
Right, I don't understand the concern here. `subspan()` would RELEASE_ASSERT if `offset` was out of bounds.

darinadler replied:
A pointer to the end of the span, past the null character, isn't out of bounds as far as the `subspan` function knows. It's in the span but we need a position that is before the last byte in the span.

kmiller68 replied:
Ah, yeah, `subspan(size())` would be a valid span but the `data()` would be invalid to dereference. It seems like it might be ok perf wise to do `RELEASE_ASSERT(!empty());` in `data()` since most(?) of the time you want the data you're about to loop over it anyway.

darinadler replied:
I realized that for this case where we are passing the length remaining in the buffer separately:
- The length in the span does us little good because it's only checking the start of the area that `vnsprintf` is writing into. We are trusting our length computation so the span isn't providing the memory safety.
- There's no reason we couldn't use `mutableSpan()` without the hull terminator instead and do the one character stricter range checking on the call to `subspan()`, but there's not much benefit to that either.

geoffreygaren replied:
FYI, I believe the full buffer hardening model, which bans memset, (not yet enforced in WebKit, but coming soon) catches this case. Specifically:

* the sprintf family of functions should warns just like memset does, since it indexes into a buffer without a bounds check

* therefore this code needs to call a vsnpritnfSpan wrapper

* if subspan() has a past the end data(), its length is zero, and vsnprintfSpan will either print zero bytes or crash, depending on our preference

We can raise this issue in our next sync with the clang team, to make sure we’re on the same page.

darinadler commented on Source/WTF/wtf/darwin/OSLogPrintStream.mm:
```diff
@@ -58,12 +58,12 @@
         m_string.grow(newLength);
         freeBytes = newLength - offset;
-        bytesWritten = vsnprintf(m_string.mutableData() + offset, freeBytes, format, backup);
+        bytesWritten = vsnprintf(m_string.mutableSpanIncludingNullTerminator().subspan(offset).data(), freeBytes, format, backup);
```
Too bad this will not detect if `offset` is past the null character.

cdumez replied:
ditto

darinadler commented on Source/WTF/wtf/posix/FileSystemPOSIX.cpp:
```diff
@@ -327,17 +327,17 @@ bool makeAllDirectories(const String& path)
-    for (; *p; ++p) {
-        if (*p == '/') {
-            *p = '\0';
+    for (; p[0]; p = p.subspan(1)) {
```
Doesn't seem like we have a strong guarantee that `p` isn't past the null terminator. Only the length check above guarantees it.

cdumez replied:
`p[0]` would RELEASE_ASSERT.

darinadler replied:
Good point. I was mistaken here.

darinadler commented on Source/WTF/wtf/text/CString.h:
```diff
@@ -44,8 +44,10 @@ class CStringBuffer final : public RefCounted<CStringBuffer> {
-    std::span<const LChar> span() const { return unsafeMakeSpan(byteCast<LChar>(data()), m_length); }
-    std::span<const char> spanIncludingNullCharacter() const { return unsafeMakeSpan(data(), m_length + 1); }
+WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN
+    std::span<const LChar> span() const { return unsafeMakeSpan(byteCast<LChar>(reinterpret_cast_ptr<const char*>(this + 1)), m_length); }
```
Why not use `LChar` in the `reinterpret_cast_ptr`?

cdumez committed:
Drop CString::mutableData(), in favor of mutableSpan()
https://bugs.webkit.org/show_bug.cgi?id=284580

Reviewed by Darin Adler.

* Source/JavaScriptCore/runtime/IntlDisplayNames.cpp:
(JSC::IntlDisplayNames::of const):
* Source/WTF/wtf/cocoa/FileSystemCocoa.mm:
(WTF::FileSystemImpl::createTemporaryZipArchive):
* Source/WTF/wtf/darwin/OSLogPrintStream.mm:
(WTF::OSLogPrintStream::vprintf):
* Source/WTF/wtf/posix/FileSystemPOSIX.cpp:
(WTF::FileSystemImpl::makeAllDirectories):
* Source/WTF/wtf/text/CString.cpp:
(WTF::CStringBuffer::createUninitialized):
(WTF::CString::copyBufferIfNeeded):
(WTF::CString::grow):
(WTF::CString::mutableData): Deleted.
* Source/WTF/wtf/text/CString.h:
(WTF::CString::data const):
(WTF::CString::spanIncludingNullTerminator const):
* Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
(WebKit::temporaryPDFDirectoryPath):
(WebKit::pathToPDFOnDisk):
* Tools/TestWebKitAPI/Tests/WTF/CString.cpp:
(TEST(WTF, CStringCopyOnWrite)):
* Tools/TestWebKitAPI/Tests/WTF/SynchronizedFixedQueue.cpp:
(TestWebKitAPI::toUpper):

Canonical link: https://commits.webkit.org/287795@main
